cmd/cloudFoundryDeploy.go

```diff
@@ -0,0 +1,855 @@
+	}
+	if len(config.Manifest) == 0 {
+		return "", fmt.Errorf("Manifest file not provided in configuration. Cannot retrieve app name")
+	}
+	fileExists, err := fileUtils.FileExists(config.Manifest)
+	if err != nil {
+		return "", err
+	}
+	if !fileExists {
+		return "", fmt.Errorf("Manifest file '%s' not found. Cannot retrieve app name", config.Manifest)
+	}
+	m, err := _getManifest(config.Manifest)
+	if err != nil {
+		return "", err
+	}
+	apps, err := m.GetApplications()
+	if err != nil {
+		return "", err
+	}
+	if len(apps) > 0 {
+		namePropertyExists, err := m.ApplicationHasProperty(0, "name")
+		if err != nil {
+			return "", err
+		}
+		if namePropertyExists {
+			appName, err := m.GetApplicationProperty(0, "name")
+			if err != nil {
+				return "", err
+			}
+			if name, ok := appName.(string); ok {
+				if len(name) > 0 {
+					return name, nil
+				}
+			}
+		}
+	}
+
+	return "", fmt.Errorf("No appName available in manifest '%s'", config.Manifest)
+}
+
+func handleSmokeTestScript(smokeTestScript string) ([]string, error) {
+
+	if smokeTestScript == "blueGreenCheckScript.sh" {
+		// what should we do if there is already a script with the given name? Should we really overwrite ...
+		err := fileUtils.FileWrite(smokeTestScript, []byte(defaultSmokeTestScript), 0755)
+		if err != nil {
+			return []string{}, err
+		}
+		log.Entry().Debugf("smoke test script '%s' has been written.", smokeTestScript)
+	}
+
+	if len(smokeTestScript) > 0 {
+		err := fileUtils.Chmod(smokeTestScript, 0755)
+		if err != nil {
+			return []string{}, err
+		}
+		pwd, err := fileUtils.Getwd()
+		if err != nil {
+			return []string{}, err
+		}
+
+		return []string{"--smoke-test", fmt.Sprintf("%s/%s", pwd, smokeTestScript)}, nil
+	}
+	return []string{}, nil
+}
+
+func prepareBlueGreenCfNativeDeploy(config *cloudFoundryDeployOptions) (string, []string, []string, error) {
+
+	smokeTest, err := handleSmokeTestScript(config.SmokeTestScript)
+	if err != nil {
+		return "", []string{}, []string{}, err
+	}
+
+	var deployOptions = []string{}
+
+	if !config.KeepOldInstance {
+		deployOptions = append(deployOptions, "--delete-old-apps")
+	}
+
+	if len(config.Manifest) > 0 {
+		manifestFileExists, err := fileUtils.FileExists(config.Manifest)
+		if err != nil {
+			return "", []string{}, []string{}, err
+		}
+
+		if !manifestFileExists {
+
+			log.Entry().Infof("Manifest file '%s' does not exist", config.Manifest)
+
+		} else {
+
+			manifestVariables, err := toStringInterfaceMap(toParameterMap(config.ManifestVariables))
+			if err != nil {
+				return "", []string{}, []string{}, err
+			}
+
+			manifestVariablesFiles, err := removeDefaultManifestVariableFileIfItIsTheOnlyFileAndThatFileDoesNotExist(config.ManifestVariablesFiles)
+			if err != nil {
+				return "", []string{}, []string{}, err
+			}
+
+			modified, err := _substitute(config.Manifest, manifestVariables, manifestVariablesFiles)
```

Comment:
`_substitute` is not a good method name.

Reply:
> _substitute is not a good method name

was more the less the same in [groovy](https://github.com/SAP/jenkins-library/blob/master/vars/cfManifestSubstituteVariables.groovy#L125). I tried to keep the names similar in order to make the shift more easy. Of course we can also improve the names. What do you suggest?

I introduced the underscore for "functions" which I would like to "replace" in the tests. Are your concerns regarding the underscore?

Reply:
I did not read this PR while comparing it to the Groovy code. To be honest, I don't plan to do so, and judging by the discussion from this morning, I don't anticipate that someone else will review this PR. And even if that were the case, if it would then be compared to the Groovy code. So I am not convinced that it is valuable to stick to the Groovy version as close as possible.

That an underscore is used is fine be me of course. Just the name "substitute" is much too generic. Even from the parameters, you couldn't guess what this does.

In general, I am in favor of doing any "obvious" improvements when migrating steps to go. To keep the code as close as possible to the Groovy version is not a priority for me. I fully understand if you see value in that to prevent regressions. That being said, I don't think improving a function name would be contrary to that objective.

All that being said, do with this suggestion as you wish, my approval certainly does not depend on it. ;-) During a review, I merely point out anything I find worth improving (and I did try to adjust my sensibility). Most often it should be clear what a "real" problem is versus what is just a suggestion. I think, "substitute" is a bad name "objectively". But it isn't a problem as in "it is broken". You have said multiple times that we can improve things in follow up PRs. And I generally agree with that. However, there is also the tendency that once something is on master, and there is no pressing need to improve upon it, then it stays as is. That's why I think there is a balance between trying to get a PR merged timely, and trying to make it as good as possible.

Reply:
> So I am not convinced that it is valuable to stick to the Groovy version as close as possible.

At least or me - not being involved to the step - it was easier to stick to the groovy code. As already outlined earlier: we should - of course - refactor (inside the step itself, but also wrt the other steps we have in the cf area in the meantime.

I changed the name to `_replaceVariables`. Hope that fits for you. If not: provide another name.